API: signal: rename `sosfreqz` to `freqz_sos`

luxedo opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Issue: https://github.com/scipy/scipy/issues/8319
Closes https://github.com/scipy/scipy/pull/8213

#### What does this implement/fix?
<!--Please explain your changes.-->
Rename `sosfreqz` to `freqz_sos` and keeps `sosfreqz` as an alias.

#### Additional information
<!--Any additional information you think is important.-->
I'd like to discuss three things:
1. Should I place `sosfreqz` in the [signa index](https://docs.scipy.org/doc/scipy/reference/signal.html)? I think that the docs build fails if it's not there.
2. I also placed a message just like the one at the end of [interpolate](https://docs.scipy.org/doc/scipy/reference/interpolate.html) for `pchip`. Is it necessary?

https://github.com/scipy/scipy/blob/776e64056f64bd70b503b7f8deb31fc155ac5c40/scipy/interpolate/__init__.py#L164-L165

3. From #8213:
> `sosfilt` `sosfilt_zi` `sosfiltfilt` should also be renamed by the same logic?

As soon as I have some feedback I wrap up the PR for merging.

luxedo committed:
MAINT: Rename `sosfreqz` to `freqz_sos`

Keep `sosfreqz` as an alias.
Co-authored-by: endolith <endolith@gmail.com>

luxedo committed:
MAINT: Add `sosfreqz` compatibility alias

larsoner commented on scipy/signal/__init__.py:
```diff
@@ -113,6 +113,11 @@
    invres        -- Inverse partial fraction expansion for analog filter.
    invresz       -- Inverse partial fraction expansion for digital filter.
    BadCoefficients  -- Warning on badly conditioned filter coefficients.
+   sosfreqz      -- Digital filter frequency response for SOS format filter.
+                    -- Alias for `freqz_sos`.
+
+``sosfreqz`` is an alias of `freqz_sos` for backward compatibility
+(should not be used in new code).
```
I don't think we need these in the module `__doc__` (which then shows up on the website). I think it's good enough to have a note in `sosfreqz`'s `__doc__` instead.

luxedo replied:
Because `sosfreqz` is not in the index, the 0.19.0 release notes can't build properly.
I'm going to try dig this a bit. Does someone knows how to get rid of this warning?
```
/home/circleci/repo/doc/source/release/0.19.0-notes.rst:117: WARNING: py:obj reference target not found: scipy.signal.sosfreqz
```

luxedo committed:
DOC: remove `sosfreqz` from signal's index

[skip actions][skip azp][skip cirrus]

larsoner commented:
You can change it from single backticks (linking mode) to double backticks (code mode) in the 0.19 release notes.

luxedo committed:
DOC: add `sosfreqz` to skiplist

luxedo commented:
Hello! Any updates on this? 

> You can change it from single backticks (linking mode) to double backticks (code mode) in the 0.19 release notes.

Yes! It is done!

larsoner commented on doc/source/release/0.19.0-notes.rst:
```diff
@@ -114,7 +114,7 @@ original signal: `scipy.signal.stft` and `scipy.signal.istft`. This
 `scipy.signal.spectrogram` when complex output data were requested.
 
-The function `scipy.signal.sosfreqz` was added to compute the frequency
+The function ``scipy.signal.sosfreqz`` was added to compute the frequency
```
Worth doing, or too confusing to cross-link?
```suggestion
The function :func:`scipy.signal.sosfreqz <scipy.signal.freqz_sos>` was added to compute the frequency
```

larsoner commented on scipy/signal/tests/test_filter_design.py:
```diff
@@ -1912,7 +1912,7 @@ def test_lowpass_1000dB(self):
         N, Wn = ellipord(wp, ws, rp, rs, False)
         sos = ellip(N, rp, rs, Wn, 'lp', False, output='sos')
-        w, h = sosfreqz(sos)
+        w, h = freqz_sos(sos)
```
Maybe leave one `sosfreqz` in here? Will get us one more line of coverage but more importantly make sure it's actually in the expected namespace for backward compat.

larsoner reviewed (commented):
Just two minor comments then I think we can merge

luxedo committed:
DOC: update `freqz_sos` release doc

Co-authored-by: Eric Larson <larson.eric.d@gmail.com>

luxedo committed:
TST: add `sosfreqz` backwards compatibility test

Co-authored-by: Eric Larson <larson.eric.d@gmail.com>

luxedo commented:
I updated the documentation with your suggestions:
https://github.com/scipy/scipy/pull/21025/files#diff-8a209fb35304df0e94887edfc436340479c2202ea7e37f7403e06135f823b093R117-R118


And I added a compatibility test similar to what's been done to other functions:
https://github.com/scipy/scipy/pull/21025/files#diff-813f319a7ef4c294752f17de96f110c7e2d8cd6cbddbe3e73f2512917ad51ac7R946-R956

I had one fail from CI but it seems it was a network error:
https://github.com/scipy/scipy/actions/runs/10675563135/job/29587743786?pr=21025

larsoner commented:
Looks good! #8213 already had an approval and but needed a bit of extra work to get it over the finish line and that has now been done (and CIs are green; I restarted the one that had a timeout), so I'll go ahead and merge. Thanks @luxedo !

luxedo commented:
Given that this is fresh in my head, I'll try to open PRs to standardize `sosfilt` `sosfilt_zi` `sosfiltfilt` as well.

@endolith in https://github.com/scipy/scipy/pull/8213#issuecomment-377970108
